### PR TITLE
Enhance IMIS build process

### LINKED ISSUE #453: SwaggerApi.ts im Build Prozess generieren
*Ausgangssituation*
Nach #452 wird SwaggerApi.ts nicht mehr in der Pipeline generiert.

*Anforderung*
Das Generieren von SwaggerApi.ts soll in den Build Prozess eingehängt werden. Der entsprechende Gradle task heißt generateClient.

### PR DESCRIPTION
**Pull request by jabrandes**
This pull request introduces various enhancements to the IMIS build process, the most significant being the integration of the complete Swagger API client generation process (closes #453) and the improvements to incremental building, which allows to skip more tasks when the respective files have not been changed (e.g. `:client:buildFrontend` and the Swagger client generation process).

**Further details on additions / modifications:**
- adds `:server:buildSwaggerConfig` build task that generates the swagger config JSON file required by the `:client:generateClient` task
- adds a replacement for the `GenerateSwagger` test: `SwaggerConfigExport`, configurable via command line arguments
- adds enhanced incremental building, reducing time to startup quite significantly when respective source code has not been touched; supported for tasks 
   - `:client:lint`
   - `:client:buildFrontend`
   - `:client:generateClient`
   - `:server:generateTestData`
   - `:server:buildSwaggerConfig`
- adds standard `run`, `build` and `jar` tasks, which should make contribution a little bit easier since working with well-known tasks
   - Note: `run` and `runFrontendDev` tasks have been added to potentially replace `:server:bootRun` and `:client:serve` tasks, respectively
- modifies the gradle inter-task dependencies such that these match up to the requirements of IMIS project
- README files are updated to reflect the changes
- CI and deployment is adopted to work with the updated build process

**Still existing problems:**
- [ ] **Test Data Loading on Cloud Run:** Since Cloud Run seems to completely restart the server after some amount of idle time, the time-costly test data insertion into the database can be triggered quite often; it would be beneficial if the database could be populated exactly once on PR deployment by a separate initialization process, which is decoupled from the IMIS server that is subject to Cloud Run process management (*restarts*)

### CODE REVIEW

**Comment by MichaelSp** [comment]
Wow! Very nice!

It doesn't work, because during CI run there is no DB available.. Is it possible to generate the swagger without a running DB?

**Comment by MichaelSp** [comment]
I think I found a way by using the In-memory H2 DB

File: client/src/views/PatientList.vue [hunk 1]
```diff
@@ -419,7 +418,9 @@ export default Vue.extend({
           formValues.patientStatus = undefined
         }
 
-        formValues.quarantineStatus = this.getQuarantineSelection() as Array<PatientStatus>
+        formValues.quarantineStatus = this.getQuarantineSelection() as Array<
+          PatientStatus
+        >
```
Comment (MichaelSp) [inline client/src/views/PatientList.vue:423]: Komisch. Bei mir wird das immer genau anders herum formatiert. Muss mal in meine settings schauen

Comment (jabrandes) [reply client/src/views/PatientList.vue:423]: Bei mir liefen nur die Tools aus dem Buildprozess die mir das hier ausgespuckt haben. An der Stelle keine IDE-Intervention.

**Comment by jabrandes** [comment]
> I think I found a way by using the In-memory H2 DB

Wouldn't surprise me if that's the database configured for Unit testing. Since `SwaggerConfigExport` is not running as such, it will look for the Postgres database instead.

**Comment by MichaelSp** [comment]
good to go, if you'd ask me

**Comment by jabrandes** [comment]
@MichaelSp I'd still like to make adjustments to the README in order to reflect the updated build process. And I think that the new tasks could use some descriptions as well, so that one knows what they do when running `gradle tasks`.

**Comment by jabrandes** [comment]
@MichaelSp There are still some bugs regarding the `jib` task. It is probably due to the resources being copied to the wrong directory. I'm on it.

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -80,7 +83,7 @@ jobs:
     # Build and push image to Google Container Registry
     - name: Build
       run: |-
-        gradle jib --image "$REGISTRY_HOSTNAME/$GKE_PROJECT_ID/imis:$GITHUB_SHA"
```
Comment (jabrandes) [inline .github/workflows/pr.yml:84]: Not sure if it was a good idea to remove the  `--image` argument. Deploy seems not that happy about it, complaining that the container image cannot be found.

Comment (MichaelSp) [reply .github/workflows/pr.yml:84]: hmm 🤔 should be covered by https://github.com/ImisDevelopers/1_011_a_infektionsfall_uebermittellung/blob/4bbbf0223e724d640f77212797c0bd44a80419fc/server/build.gradle#L81 and works locally ... 🤔

Comment (jabrandes) [reply .github/workflows/pr.yml:84]: Problem is that this image name is not exactly equal to what the image of the reviewed code line was named previously (note the `imis:` vs. the `imis-server:`). And Deploy expects exactly that name.

**Comment by jabrandes** [comment]
@MichaelSp The last thing that needs to be fixed before merging: Logging into the deployed image does not work, which is most likely due to the test data not being loaded. Could you make sure that the deployed image is run with the `feature.testdata` property set to `true`? You just have to make sure that test data is not loaded during Swagger generation, since that will throw an exception.

**Comment by MichaelSp** [comment]
Login currently not working. Reason unclear. Potential issues:
 * DB not fully populated?!
 * CORS issues?!

**Comment by jabrandes** [comment]
Logging into the deployed image is now working correctly, since CORS is no longer handled by Spring and the very long-running process of inserting patient test data is executed AFTER the test accounts, so that one can now log in after a reasonable amount of time. However, due to CORS being disabled, we now got the problem that the client code served by the frontend development server (alas `runFrontendDev`) cannot issue API calls :roll_eyes:.

**Comment by MichaelSp** [comment]
Isn't https://github.com/ImisDevelopers/1_011_a_infektionsfall_uebermittellung/blob/9e7e0d97f4741eadb6e7d1e50a2b5bb55aa5ec0a/server/src/main/resources/application.yml#L72 and/or https://github.com/ImisDevelopers/1_011_a_infektionsfall_uebermittellung/blob/9e7e0d97f4741eadb6e7d1e50a2b5bb55aa5ec0a/server/src/main/resources/application.yml#L77 the problem? There is no proper schema evolution. I'm not a spring expert, but shouldn't we use Flyway or Liquibase?

**Comment by jabrandes** [comment]
@MichaelSp In the long run definitely, since it would be no good to just drop the database on an update :stuck_out_tongue:. I think for testing it would be nice if we just had a separate component that initializes the database that could be executed independently of the actual server system.

**Comment by MichaelSp** [comment]
hm, so H2 in memory for PR?!

**Comment by jabrandes** [comment]
I know. I just tested what happens when executing all those patient insertions in a single transaction, hoping for the persistence layer to batch the operations and speed up the process. But then any patient data is only available after the whole transaction, which still takes its time. 

In my opinion, any further database usage optimization could be content for another issue. I think the current state is stable enough that it can be considered merge-able.

**Pull request merged**